pgjdbc/src/test/java/org/postgresql/test/jdbc4/PGCopyInputStreamTest.java

```diff
@@ -5,111 +5,112 @@
+  private static final String CREATE_TABLE_SQL = String.format(
+      "CREATE TABLE %s AS SELECT i FROM generate_series(0, %d) i", TEST_TABLE, NUM_TEST_ROWS - 1);
+  private static final String COPY_SQL = String.format("COPY %s TO STDOUT", TEST_TABLE);
+  private static final String COPY_IN_ORDER_SQL = String
+      .format("COPY (SELECT i FROM copy_input_stream_test ORDER BY i ASC) TO STDOUT", TEST_TABLE);
+
   private Connection conn;
-  private PGCopyInputStream sut;
-  private String copyParams;
 
   @Before
   public void setUp() throws Exception {
     conn = TestUtil.openDB();
-    TestUtil.createTable(conn, "cpinstreamtest", "i int");
-    if (TestUtil.haveMinimumServerVersion(conn, ServerVersion.v9_0)) {
-      copyParams = "(FORMAT CSV, HEADER false)";
-    } else {
-      copyParams = "CSV";
-    }
+    TestUtil.dropTable(conn, TEST_TABLE);
+    TestUtil.execute(CREATE_TABLE_SQL, conn);
   }
 
   @After
   public void tearDown() throws SQLException {
-    silentlyCloseStream(sut);
-    TestUtil.dropTable(conn, "cpinstreamtest");
+    TestUtil.dropTable(conn, TEST_TABLE);
     TestUtil.closeDB(conn);
   }
 
   @Test
   public void testReadBytesCorrectlyHandlesEof() throws SQLException, IOException {
-    insertSomeData();
-
-    sut = new PGCopyInputStream((PGConnection) conn,
-        "COPY cpinstreamtest (i) TO STDOUT WITH " + copyParams);
-
-    byte[] buf = new byte[100]; // large enough to read everything on the next step
-    assertTrue(sut.read(buf) > 0);
-
-    assertEquals(-1, sut.read(buf));
+    PGConnection pgConn = conn.unwrap(PGConnection.class);
+    try (PGCopyInputStream in = new PGCopyInputStream(pgConn, COPY_SQL)) {
+      // large enough to read everything on the next step
+      byte[] buf = new byte[COPY_DATA_SIZE + 100];
+      assertEquals("First read should get the entire table into the byte array",
+          COPY_DATA_SIZE, in.read(buf));
+      assertEquals("Subsequent read should return -1 to indicate stream is finished",
+          -1, in.read(buf));
+    }
   }
 
   @Test
   public void testReadBytesCorrectlyReadsDataInChunks() throws SQLException, IOException {
-    insertSomeData();
-
-    sut = new PGCopyInputStream((PGConnection) conn,
-        "COPY (select i from cpinstreamtest order by i asc) TO STDOUT WITH " + copyParams);
-
-    byte[] buf = new byte[2]; // small enough to read in multiple chunks
-    StringBuilder result = new StringBuilder(100);
-    int chunks = 0;
-    while (sut.read(buf) > 0) {
-      result.append(new String(buf));
-      ++chunks;
+    List<String> chunks = new ArrayList<>();
+    PGConnection pgConn = conn.unwrap(PGConnection.class);
+    try (PGCopyInputStream in = new PGCopyInputStream(pgConn, COPY_IN_ORDER_SQL)) {
+      // Allocate size of exactly one row
+      byte[] buf = new byte[COPY_ROW_SIZE];
+      while (in.read(buf) > 0) {
+        chunks.add(new String(buf, StandardCharsets.UTF_8));
```

Comment:
What are you suggesting I replace it with here? The byte sequence is expected to exactly match the entirety of the string so I'm don't see how it could get anything else in response. Even if it did, it's checking for an exact match after anyway.

Reply:
E.g. add `readFully(InputStream)` helper method and convert everything into `String` at once.

Reply:
Ah I didn't do that because I wanted to count the chunks as well. It's not just testing that the totality of it creates the correct output, but that we had to read N (or N -1) chunks.

Reply:
Then collect chunks to `List<byte[]>`, join them, and then convert to string.